etl/steps/data/garden/hmd/2022-12-07/life_tables.py

```diff
@@ -0,0 +1,136 @@
+
+    # build tables
+    for table_new, tables_old in TABLE_MAPPING.items():
+        log.info(f"life_tables: {table_new}.start")
+        tb_garden = make_table(ds_meadow, tables_old, table_new)
+        ds_garden.add(tb_garden)
+        ds_garden.save()
+        log.info(f"life_tables: {table_new}.end")
+
+    log.info("life_tables.end")
+
+
+def make_table(ds_meadow: Dataset, tables_old_names: List[str], table_new_name: str) -> Table:
+    """Create a Garden table from multiple Meadow tables.
+
+    In Meadow, we have a table per agexyear and sex. In Garden, brin the sex dimension in the table as
+    a column in the index. That is, tables "both_1x1", "female_1x1", "male_1x1" are combined into one table
+    called "period_1x1".
+    """
+    log.info(f"life_tables: building table {table_new_name} from {tables_old_names}...")
+    # Combine multiple tables (broken down by sex) into single one (insert sex in index)
+    df = combine_sex_tables(ds_meadow, tables_old_names)
+
+    # Country management
+    df = clean_countries(df)
+
+    # Build table
+    tb_garden = underscore_table(Table(df))
+
+    # Edit table
+    tb_garden.update_metadata_from_yaml(N.metadata_path, table_new_name)
+    tb_garden = tb_garden.set_index(["country", "year", "age", "sex"], verify_integrity=True).sort_index()
+
+    return tb_garden
+
+
+def combine_sex_tables(ds_meadow: Dataset, table_names: List[str]) -> pd.DataFrame:
+    """Combine meadow tables."""
+    dfs = []
+    for table_name in table_names:
+        tb_meadow = ds_meadow[table_name]
+        # Country management
+        df = pd.DataFrame(tb_meadow)
+        # Sex
+        if table_name.startswith("both_"):
+            df = df.assign(sex="all")
+        elif table_name.startswith("female_"):
+            df = df.assign(sex="female")
+        elif table_name.startswith("male_"):
+            df = df.assign(sex="male")
+        else:
+            raise ValueError(f"Unknown sex for table {table_name}!")
+        dfs.append(df)
+    df = pd.concat(dfs, ignore_index=True)
+    return df  # type: ignore
+
+
+def clean_countries(df: pd.DataFrame) -> pd.DataFrame:
+    """Clean country names, and discard those countries not used."""
+    df = exclude_countries(df)
+    df = harmonize_countries(df)
+    return df
+
+
+def load_excluded_countries() -> List[str]:
+    """Load list of excluded countries from JSON file."""
+    with open(N.excluded_countries_path, "r") as f:
+        data = json.load(f)
+        assert isinstance(data, list)
+    return data
+
+
+def exclude_countries(df: pd.DataFrame) -> pd.DataFrame:
+    """Exclude countries."""
+    excluded_countries = load_excluded_countries()
+    return cast(pd.DataFrame, df.loc[~df.country.isin(excluded_countries)])
+
+
+def harmonize_countries(df: pd.DataFrame) -> pd.DataFrame:
+    """Country harmonization."""
+    unharmonized_countries = df["country"]
+    df = geo.harmonize_countries(df=df, countries_file=str(N.country_mapping_path))
```

Comment:
Even if we know that harmonize_countries needs some additional work, an easy win would be to add the option to pass the path to the excluded countries file. Having that we could spare having to define exclude_countries, and harmonize_countries every time.

Reply:
Good idea, moved this to https://github.com/owid/owid-datautils-py/issues/55